Preliminary support for UI Automation in Windows Console

codeofdusk opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:
This PR shouldn't yet close any issues, but it is related to #513, #673, #1682, #6291, and #7497.
### Summary of the issue:
NVDA provides support for the Windows command console used by Command Prompt, PowerShell, and the Windows Subsystem for Linux. It allows users to read the console's contents using review cursor commands, and automatically reports console output as it arrives. While functional, this support has some shortcomings, leading to decreased NVDA performance and stability and, by extension, user productivity.

### Description of how this pull request fixes the issue:
This PR re-implements NVDA's Windows Console support using Microsoft's UI Automation API (UIA). At the moment, typed characters and words are read to the user as well as incoming text. Review of console output is functional on Windows 10 1803 and 1809. The new console support is implemented in the `NVDAObjects.UIA.winConsoleUIA.winConsoleUIA` object, and can be enabled with a new option in NVDA's advanced preferences.

Since the UIA console interface was first implemented in Windows 10 version 1709, it is safe to use the `keyboardHandler` module to receive typed characters in UIA consoles. The console sometimes sends duplicate `WM_CHAR` messages for typed characters, so the in-process approach makes "speak typed words" unreliable.

On windows 10 version 1903 and later, the UIA caret position is off-by-one. To compensate, `consoleUIATextInfo` moves new UIA text ranges around the caret one character to the right.

A mapping from UIA's event for text changes to NVDA's `textChange` event was present, but commented out, in `_UIAHandler`. This mapping has been uncommented to allow automatic readout of console text to be implemented.

Automatic text readout uses the `NVDAObjects.behaviors.LiveText` functionality. Since UIA adds thousands of empty lines to console output, it was necessary to re-implement `_getTextLines` to filter the output.

UIA fires `textChange` whenever the console's text changes, even during user input. While `LiveText` already filters out many typed characters, some duplicates (particularly the beginning of new words) were being reported. To compensate, `winConsoleUIA` does not report any new text while the user is typing. This is not a regression of NVDA's functionality, as speech is already cancelled during user input, even when "speak typed characters" and "speak typed words" are disabled.
### Testing performed:
Tested the new support on Windows 10 versions 1803, 1903, and 2003. In particular, tested reporting of typed characters and words (when the corresponding settings were enabled), reporting of new text, text editing and selection, and object review.
### Known issues with pull request:

* In object review, commands for moving to and reporting words move to or report lines instead.
* A large number (thousands in many cases) of empty lines are appended to the console's object review.
* Reporting of text selection is incorrect on Windows 10 version 1809 and earlier.
* Sometimes, automatic readout begins reading the terminal window from the top instead of reporting only new text (I think this is related to printing empty lines or clearing the screen, it can be reproduced when closing the nano text editor). Maybe fixed in 9d43eb0eb258c0adbab348a35406f4290d75f35c.
* When "speak typed characters" or "speak typed words" is enabled, typed text that is not displayed onscreen (such as passwords) is still reported.
* After all characters of the input line have been deleted, further presses of backspace read the last character of the output line.
* When reviewing to the last line of the console, the review cursor may become stuck. When this happens, it is impossible to review text in the console until moving focus away from the window and back again.
* Sometimes, attempting to interact with the console too quickly after focusing the window can make NVDA unable to get any caret or review information or receive new text events. To work around this, move focus away from the window and back again.
* In console applications that update the screen while the user is typing (i.e. before `winConsoleUIA._isTyping` is cleared), reporting of new text and speak typed words may not function correctly.
* If the advanced setting "force UI Automation in the Windows Console" is enabled on a system without a UIA console (such as Windows versions below Windows 10 1709 or systems with the "use legacy console" option selected), the console is completely inaccessible.
* If "speak typed words" is enabled, the last typed word is announced when pressing enter.
* When pressing an interrupt character (such as control+c), the `_isTyping` flag is not cleared on the console, so new text is not announced immediately.

### Change log entry:

== New features ==
- Added an experimental option to the Advanced Settings panel that allows you to try out a new, work-in-progress rewrite of NVDA's Windows Console support using the Microsoft UI Automation API. (#9614)

== Changes for Developers ==
- Added a new `isAtLeastWin10` function to the `winVersion` module which returns whether or not this copy of NVDA is running on at least the supplied release version (such as 1809 or 1903) of Windows 10. (#9614)

codeofdusk committed:
Add a feature flag for UIA console support.

At the moment, it merely disables NVDA's existing console support, falling back to what UIA support is available in NVDA. The option is currently called `ConsoleUIA`, but we'll need to change this once autodetection is implemented (as there will be three options at that point).

Typed characters are now spoken in UIA consoles, but characters are doubled when typing quickly, leading to incorrect typed word reporting.

codeofdusk committed:
Disable in-process injection for speak typed characters for now.

This fixes the doubled characters bug, making speaked typed words function as intended, but it may have other implications.

codeofdusk committed:
Receive textChanged events from UIA and add consoleUIA and consoleUIATextInfo classes to implement automatic readout and work around a caret movement bug in Windows 10 version 1903.

The consoleUIA class inherits from NVDAObjects.behaviors.Terminal, which serves as the foundation for automatic readout support. At the moment, it is likely that such support will either be not functional at all or report extraneous text from user input.

codeofdusk committed:
Introduce basic automatic readout support, and rename ConsoleUIATextInfo->consoleUIATextInfo for consistency.

See the source code comments in this commit for some steps still incomplete.

codeofdusk committed:
Filter out extraneous characters by disabling live text reporting while a user is typing.

This does not regress NVDA functionality, since we already cancel speech when the user types.

codeofdusk committed:
Cleanup.

codeofdusk committed:
Style, eliminate magic number, halve typing timeout, update label in GUI to inform the user that restarting NVDA is required to switch console implementations.

codeofdusk committed:
Merge branch 'master' into cmduia1

codeofdusk committed:
Re-implement consoleUIA._getTextLines using UIATextPattern.GetVisibleRanges. This may improve performance when large amounts of text are written to the console, but possibly at the cost of less reliable autoread (particularly when the screen is refreshed or scrolled).

codeofdusk committed:
Clear isTyping when tab is pressed to announce tab completion.

feerrenrut commented on source/NVDAObjects/UIA/ConsoleUIA.py:
```diff
@@ -0,0 +1,64 @@
+		self.event_textChange()
+
+	def event_typedCharacter(self, ch):
+		if len(''.join(ch.split())) > 0:
```
why split then join here?

codeofdusk replied:
To eliminate whitespace.

codeofdusk replied:
I like `isspace`. Thanks!

feerrenrut commented on source/gui/settingsDialogs.py:
```diff
@@ -2050,6 +2050,13 @@ def __init__(self, parent):
 
+		# Translators: This is the label for a checkbox in the
+		#  Advanced settings panel.
+		label = _("Use UI Automation to access the Windows Console (highly experimental, restart NVDA for changes to take effect)")
```
I think you can leave out the warning. The settings panel has quite a warning on it already. You can add a similar message to the user guide though.
```suggestion
		label = _("Use UI Automation to access the Windows Console.")
```

feerrenrut commented on source/NVDAObjects/UIA/ConsoleUIA.py:
```diff
@@ -0,0 +1,64 @@
+		# Todo: do this (also) somewhere else so they aren't in document review either
+		ptr = self.UIATextPattern.GetVisibleRanges()
+		res = [ptr.GetElement(i).GetText(-1) for i in range(ptr.length)]
+		return res
```
Missing a newline at the end of the file.

feerrenrut reviewed (commented):
A few small things, overall it looks good! Could you also update the user guide to mention the new setting too.

codeofdusk committed:
Review actions, style, update user guide.

The user guide will need to be updated once auto-detection is introduced (and therefore the NVDA preference is changed).

leonardder commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -858,6 +858,10 @@ def findOverlayClasses(self,clsList):
 			pass
 
+		# Support Windows Console's UIA interface
+		if self.windowClassName == "ConsoleWindowClass" and self.UIAElement.cachedAutomationId == "Text Area" and config.conf['UIA']['consoleUIA']:
```
Please split this:

```suggestion
		if (
			self.windowClassName == "ConsoleWindowClass"
			and self.UIAElement.cachedAutomationId == "Text Area"
			and config.conf['UIA']['consoleUIA']
		):
```

leonardder replied:
I guess the config check is not necessary here, as I think the badUIAWindowClassNames check already deals with this.

leonardder commented on source/_UIAHandler.py:
```diff
@@ -59,13 +59,15 @@
 
+# #7497: Windows 10 Fall Creators Update has an incomplete UIA implementation for console windows, therefore for now we should ignore it.
+# It does not implement caret/selection, and probably has no new text events.
+if not config.conf['UIA']['consoleUIA']:
```
I assume you will extend this check in the future? On windows 7, 8 and 10 below fall creators update, this will break in a major way probably.

codeofdusk replied:
Yes – the final PR for this work will introduce auto-detection for UIA/legacy console support which will make checking more robust as a consequence.

leonardder commented on source/_UIAHandler.py:
```diff
@@ -140,7 +142,7 @@
 UIAEventIdsToNVDAEventNames={
 	UIA_LiveRegionChangedEventId:"liveRegionChange",
-	#UIA_Text_TextChangedEventId:"textChanged",
+	UIA_Text_TextChangedEventId:"textChanged",
```
Shouldn't this be:

```suggestion
	UIA_Text_TextChangedEventId:"textChange",
```

This makes the event_textChanged definition in UIAConsole obsolete.

Alternatively, if you don't want to map this to the textChange event, it might be better to prefix the event name with UIA, i.e:

```suggestion
	UIA_Text_TextChangedEventId:"UIA_textChanged",
```

codeofdusk replied:
This line was commented out, and I didn't want to change it (thought it was called `textChanged` for a reason). I agree that it's confusing though!

A quick scan of the NVDA source shows that `textChange` is used more than `textChanged` (in fact, `textChanged` is basically only used for UIA). Surely this must've been a mistake.

leonardder commented on source/config/configSpec.py:
```diff
@@ -187,6 +187,7 @@
 [UIA]
 	enabled = boolean(default=true)
 	useInMSWordWhenAvailable = boolean(default=false)
+	consoleUIA = boolean(default=false)
```
```suggestion
	useInConsoleWhenAvailable = boolean(default=false)
```

codeofdusk replied:
This setting is only temporary and "use when available" is misleading (as that isn't what it does). The final PR will change the setting to something like:

    winConsoleImplementation= option("auto", "legacy", "UIA", default="auto")

leonardder commented on source/gui/settingsDialogs.py:
```diff
@@ -2050,6 +2050,13 @@ def __init__(self, parent):
 
+		# Translators: This is the label for a checkbox in the
+		#  Advanced settings panel.
+		label = _("Use UI Automation to access the Windows Console (restart NVDA for changes to take effect)")
```
Why is a restart necessary?

codeofdusk replied:
Changing the setting seems to leave NVDA in an inconsistent state until it's restarted. Maybe because of the `badUIAWindowClassNames` insertion?

leonardder commented on source/keyboardHandler.py:
```diff
@@ -205,10 +205,12 @@ def internal_keyDownEvent(vkCode,scanCode,extended,injected):
 			and not isNVDAModifierKey(vkCode,extended) and not vkCode in KeyboardInputGesture.NORMAL_MODIFIER_KEYS
 			and ( # Either of
-				# We couldn't inject in-process, and its not a console window (console windows have their own specific typed character support)
+				# We couldn't inject in-process, and its not a legacy console window (console windows have their own specific typed character support)
```
Split:

```suggestion
				# We couldn't inject in-process, and its not a legacy console window.
				# console windows have their own specific typed character support.
```

leonardder commented on source/keyboardHandler.py:
```diff
@@ -205,10 +205,12 @@ def internal_keyDownEvent(vkCode,scanCode,extended,injected):
 				# or the focus is within a UWP app, where WM_CHAR never gets sent 
 				or focus.windowClassName.startswith('Windows.UI.Core')
+				#Or this is a UIA console window, where WM_CHAR messages are doubled
+				or focus.windowClassName == "ConsoleWindowClass" and config.conf['UIA']['consoleUIA']
```
I don't believe this is the right way to ensure that UIA is enabled, especially not for older versions of Windows. I think I'd rather see an instance check of the object

LeonarddeR reviewed (changes requested):
This really looks promising.

codeofdusk commented on source/NVDAObjects/UIA/ConsoleUIA.py:
```diff
@@ -0,0 +1,64 @@
+				# The UIA implementation in 1903 causes the caret to be
+				# off-by-one, so move it one position to the right
+				# to compensate.
+				self._rangeObj.MoveEndpointByUnit(
```
The `move` method is pretty broken at the moment on 1903 (at least in the context of document review). Will experiment with this for the next PR.

codeofdusk commented on source/NVDAObjects/UIA/ConsoleUIA.py:
```diff
@@ -0,0 +1,64 @@
+			return
+		super(consoleUIA, self)._reportNewText(line)
+
+	def event_textChanged(self):
```
See below.

codeofdusk committed:
Review actions.

codeofdusk commented:
@feerrenrut @leonardder I think I've addressed all review items. Could you please have another look?

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -0,0 +1,64 @@
+			return
+		super(winConsoleUIA, self)._reportNewText(line)
+
+	def event_textChanged(self):
```
This is now obsolete.

codeofdusk replied:
Oops, thought I deleted that last time...

leonardder commented on source/_UIAHandler.py:
```diff
@@ -140,7 +142,7 @@
 UIAEventIdsToNVDAEventNames={
 	UIA_LiveRegionChangedEventId:"liveRegionChange",
-	#UIA_Text_TextChangedEventId:"textChanged",
+	UIA_Text_TextChangedEventId:"textChange",
```
Having thought some more about this... Note that there was a good reason in the past to comment this out, see https://github.com/nvaccess/nvda/commit/9025c04dc0aaef7fc5ae4575844260503b5acdc5 and issue #1205.

We should consider this carefully.

michaelDCurran replied:
At very least, we should only register for this event on Windows 10 and above. I don't think we can  hold back because of .net 3.5 for ever, as these are very useful events. Plus, Windows 10 now has improved UIA event filtering which should hopefully  solve the issue on windows 10.

leonardder commented on source/config/configSpec.py:
```diff
@@ -187,6 +187,7 @@
 [UIA]
 	enabled = boolean(default=true)
 	useInMSWordWhenAvailable = boolean(default=false)
+	consoleUIA = boolean(default=false)
```
> This setting is only temporary and "use when available" is misleading (as that isn't what it does). The final PR will change the setting to something like:
> `winConsoleImplementation= option("auto", "legacy", "UIA", default="auto")`

Mmmm, honestly I don't see why `use when available` is misleading. In fact, the current approach allows every windows 7 or windows 8.x user to enable this option without any warning, likely getting major issues in command consoles on these versions. Even when this option is hidden in the advanced category, I really prefer a firmer constraint.

If I understand you correctly:

* Auto is supposed to be UIA if the Windows 10 version supports it properly
* UIA always uses UIA, regardless what Windows version is in use
* Legacy always uses legacy

In this case, auto can be explained as `use UIA when available and working OK`. This is similar to how the useInMSWord boolean works. Again, we really should avoid using UIA for command consoles in version of Windows where UIA is most definitely not supported for them.

Please also note that changing the option from boolean to option would require a config spec change and a config sscheme upgrade. I'd really prefer if we could reach consensus about something for the config spec that could be persistent from the start.

codeofdusk replied:
Ah, I probably should've clarified a few things:

* Windows 10 has had UIA console support since version 1709, but it was incomplete in early builds (see #7497). The final PR will set a minimum good Windows version for UIA, which likely won't be 1709 (even though 1709 has *limited* support). That is to say, we won't turn on UIA console support automatically until a Windows version where we know it works completely, but we should give users the option of forcibly enabling it if our default is too aggressive.
* I think that @feerrenrut and @michaelDCurran were planning on incrementally merging UIA console support into `master` during development (similar to UIA word support). We obviously don't want to enable it by default during development (so we don't break production systems), but once it's ready we want to enable it for all users (unless they explicitly turn it off). TO accomplish this, I proposed the provisional setting during development (and then adding the auto-detection and new setting once it's final), but maybe there's a better way?

michaelDCurran replied:
Checkbox checked is mapped to UIA. Checkbox unchecked is mapped to auto. The label of the checkbox now has the word "force" in it.

codeofdusk committed:
Remove unneeded event_textChanged

codeofdusk committed:
Fix document review on Windows 10 1903.

This fix adds a new textInfo instance variable, _expandCollapseBeforeReview, which stops review from expanding and immediately collapsing the textInfo during review. For some reason, this was causing object review to malfunction in this instance.

codeofdusk committed:
Remove 1903 warning from user guide.

michaelDCurran commented on source/_UIAHandler.py:
```diff
@@ -59,13 +59,15 @@
 
+# #7497: Windows 10 Fall Creators Update has an incomplete UIA implementation for console windows, therefore for now we should ignore it.
+# It does not implement caret/selection, and probably has no new text events.
+if not config.conf['UIA']['consoleUIA']:
```
It would be good if this could be checked at runtime so that changing the option in the Advanced settings panel allows UIA to be used in consoles without restarting NVDA. I would suggest adding an isBadUIAWindow function to _UIAHandler, which should check badUIAWindowClassNames and also use this code. Then in _isUIAWindowHelper, isBadUIAWindow should be called instead of checking badUIAWindowClassNames directly.

michaelDCurran commented on source/winConsoleHandler.py:
```diff
@@ -46,7 +46,6 @@
 COMMON_LVB_UNDERSCORE=0x8000
 
 
-
```
Looks like removing this blank line is the only change in this file. This change can probably be removed.

michaelDCurran commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -0,0 +1,62 @@
+				# to compensate.
+				self._rangeObj.MoveEndpointByUnit(
+					UIAHandler.TextPatternRangeEndpoint_Start,
+					UIAHandler.NVDAUnitsToUIAUnits['character'],
```
Can you please use textInfos.UNIT_CHARACTER rather than the literal string here.

codeofdusk committed:
Check bad UIA window class name at runtime.

This allows for UIA console support to be dynamically enabled/disabled.

Note: _UIAHandler.badUIAWindowClassNames has been removed. To check if a UIA windowClassName is bad, use the new _UIAHandler.UIAHandler._isBadUIAWindowClassName method.

codeofdusk committed:
Add a new winVersion.isAtLeastWin10 function for checking win10 builds with friendly version numbers, and only register for UIA textChange on Windows 10.

codeofdusk committed:
Misc review actions.

codeofdusk commented:
Could you please have another look @michaelDCurran ?

michaelDCurran commented on source/_UIAHandler.py:
```diff
@@ -344,6 +323,32 @@ def IUIAutomationNotificationEventHandler_HandleNotificationEvent(self,sender,No
+		# It does not implement caret/selection, and probably has no new text events.
+		if not config.conf['UIA']['consoleUIA']:
+			badUIAWindowClassNames.append("ConsoleWindowClass")
+		return windowClass in badUIAWindowClassNames
```
If I am reading this correctly, this will  continue adding consoleWindowClass to badUIAWindowClassNames over and over again. To avoid this, and to make the diff for this pr more readable I would suggest that badUIAWindowClassNames  stay where it originally was, and then in this function, if the window class is ConsoleWindowClass and  UIA in consoles is configured, you return True. If not, you check if windowClass is in badUIAWindowClassNames. I.e. ConsoleWindowClass is never cached.

codeofdusk replied:
Could `badUIAWindowClassNames` be accessed from somewhere else?

michaelDCurran replied:
It could. We could address this by renaming badUIAWindowClassNames to _badUIAWindowClassNames denoting that this variable is private to the module. But I'm not sure if it is all that important.

leonardder commented on source/_UIAHandler.py:
```diff
@@ -154,6 +132,9 @@
 	UIA_SystemAlertEventId:"UIA_systemAlert",
 }
 
+if winVersion.isAtLeastWin10():
```
While I really like this new winVersion helper function, I think it might be better to tie this to a minimum version of the UIAutomation object. This also makes sure that it works in source copies of NVDA. If I'm correct, Windows 10 started with UIAutomation4, but @josephsl may no more.

I think we could also reregister for UIA_Text_TextSelectionChangedEvent while at it.

michaelDCurran replied:
As much as linking this to UIAutomation interfaces would make this work for source copies,  it doesn't quite fit logically, nor is there necessarily an interface for the exact release version we want to match against.

leonardder commented on source/winVersion.py:
```diff
@@ -26,3 +26,27 @@ def canRunVc2010Builds():
+	try:
+		return winVersion.build >= win10VersionsToBuilds[version]
+	except KeyError:
+		log.warning("Unknown Windows 10 version {}".format(version))
```
I'd rather see this raise an exception, but good to know opinions from others as well here.

michaelDCurran replied:
Change this to log.error please. There is no time where it would be okay for this error to be ignored.

leonardder commented on source/winVersion.py:
```diff
@@ -26,3 +26,27 @@ def canRunVc2010Builds():
+		return winVersion.build >= win10VersionsToBuilds[version]
+	except KeyError:
+		log.warning("Unknown Windows 10 version {}".format(version))
+		return False
```
Please add an empty line below this line.

michaelDCurran commented on source/winVersion.py:
```diff
@@ -26,3 +26,27 @@ def canRunVc2010Builds():
+def isAtLeastWin10(version=1507):
+	"""
+	Returns True if NVDA is running on at least the supplied version of Windows 10. If no argument is supplied, returns True for all public Windows 10 releases.
+	Note: this function will always return False for source copies of NVDA due to a Python bug.
```
Can you specifically explain the 'version' argument in the docstring, in that this is a public release version and not a build version?

LeonarddeR reviewed (commented):
Arrg, really frustrating that GitHub calls every way to comment on an ongoing conversation for an inline comment that is outdated. I'm taking the discussion about the config spec to the main threat here.

> Ah, I probably should've clarified a few things:
> + Windows 10 has had UIA console support since version 1709, but it was incomplete in early builds (see #7497). The final PR will set a minimum good Windows version for UIA, which likely won't be 1709 (even though 1709 has limited support). That is to say, we won't turn on UIA console support automatically until a Windows version where we know it works completely, but we should give users the option of forcibly enabling it if our default is too aggressive.

I agree that this makes sense.

> + To accomplish this, I proposed the provisional setting during development (and then adding the auto-detection and new setting once it's final), but maybe there's a better way?

As noted, a provisional setting requires a config spec and scheme update as soon as it gets changed. Especially if you already know what the final setting is going to look like, I'd prefer just adding the setting to the spec as it is supposed to be in the final pr. You could map UIA to True and legacy to False and just make auto act like False/legacy for now.

codeofdusk committed:
Review actions.

codeofdusk committed:
Change config spec

leonardder commented:
@michaeldcurran: Would be good to know what your opinion is about my concerns with the current configuration parameter for this feature. See https://github.com/nvaccess/nvda/pull/9614#pullrequestreview-242106626

codeofdusk commented:
> @michaelDCurran: Would be good to know what your opinion is about my concerns with the current configuration parameter for this feature. See [#9614 (review)](https://github.com/nvaccess/nvda/pull/9614#pullrequestreview-242106626)

Addressed (I think). My only concern is what happens if users explicitly disable the incomplete feature during development (that pref won't be overridden once it's final, which it should be at least once).

josephsl commented:
Hi,

To Bill, congratulations on achieving this milestone.

As for UIA and Windows 10: it began its life with UIA3, inheriting things from Windows 8.1. Version 1607 introduced UIA4 (UIA changes event was introduced), followed by UIA5 in Version 1709 (the famous UIA notification event). Version 1809 introduced UIA6, which among other things, introduced dialog detection, event groups, and coalescing. UIA6 did help us resolve long-standing performance bugs, but Mick spent days writing additional fixes for certain cases.

Thanks.

codeofdusk commented:
@leonardder The presence of the old setting in the config doesn't seem to break anything, and the checkbox mapping to legacy/UIA is a bit confusing (at least to me, it makes the code a little less readable). Why shouldn't we hold off on changing the setting from `consoleUIA` to `winConsoleImplementation` until the final PR? Does this break installed copies?

codeofdusk committed:
Meeting actions.

codeofdusk committed:
Make auto the default.

codeofdusk committed:
Use->force

codeofdusk commented:
@leonardder @michaelDCurran @feerrenrut OK, I think this PR is now complete. I won't merge further console UIA work into this branch so this PR can stand.

codeofdusk committed:
Merge branch 'master' into cmduia

michaelDCurran committed:
Updated what's new.

michaelDCurran committed:
Settings dialog: Ensure that the UIA in Windows Consoles option can be restored to default using the Advance Panel's restore to defaults button.

codeofdusk commented:
Sorry, small error in what's new:

== Changes for Developers ==
- Added a new isAtLeastWin10 function to the winVersion module which returns whether or not this copy of NVDA is running on at least the supplied release version of Windows 10 (such as 1809 or 1903) of Windows 10. (#9614)

Should be:
== Changes for Developers ==
- Added a new isAtLeastWin10 function to the winVersion module which returns whether or not this copy of NVDA is running on at least the supplied release version (such as 1809 or 1903) of Windows 10. (#9614)

beqabeqa473 commented:
hello.

when using this new feature, i discovered that navigation by 4/6 through words doesn't work.
navigating with 7/9 and 4/6 on numpad in identical.

codeofdusk commented:
@beqabeqa473 This is a known issue as stated above. Partially addressed in #9646.

codeofdusk commented:
@beqabeqa473 Fixed in #9647.
